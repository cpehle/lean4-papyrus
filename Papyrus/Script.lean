import Papyrus.Script.Module
import Papyrus.Script.Function
import Papyrus.Script.Instructions
import Papyrus.Script.Types
