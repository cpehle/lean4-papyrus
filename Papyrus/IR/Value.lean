import Papyrus.Context
import Papyrus.IR.TypeRef

namespace Papyrus

/--
  An external reference to the LLVM representation of a Value.
  See https://llvm.org/doxygen/classllvm_1_1Value.html.
-/
constant ValueRef : Type := Unit

namespace ValueRef

/-- Get a reference to this value's type. -/
@[extern "papyrus_value_get_type"]
constant getType (self : @& ValueRef) : IO TypeRef

/-- Get whether this value has a name. -/
@[extern "papyrus_value_has_name"]
constant hasName (self : @& ValueRef) : IO Bool

/-- Get the name of this value (or the empty string if none). -/
@[extern "papyrus_value_get_name"]
constant getName (self : @& ValueRef) : IO String

/--
  Set the name of this value.
  Passing the empty string will remove the value's name.
-/
@[extern "papyrus_value_set_name"]
constant setName (name : @& String) (self : @& ValueRef) : IO PUnit

/-- Print the IR of this value to standard error for debugging. -/
@[extern "papyrus_value_dump"]
constant dump (self : @& ValueRef) : IO PUnit

end ValueRef

/--
  An external reference to the LLVM representation of a User.
  See https://llvm.org/doxygen/classllvm_1_1User.html.
-/
def UserRef := ValueRef
