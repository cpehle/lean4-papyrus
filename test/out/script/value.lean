import Papyrus.Script.Value

open Papyrus Script

#check show ModuleM PUnit from do
  let x ← llvm false
  let x ← llvm true
  let x ← llvm i32 0
  let x ← llvm i64 -1
  let x ← llvm "hello"*
  let x ← llvm "hello" addrspace(5) *
  let x ← llvm %x
  pure ()
