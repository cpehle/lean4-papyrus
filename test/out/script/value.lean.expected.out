(fun this => this)
  do 
    let _ ← liftM ConstantIntRef.getFalse 
    let _ ← liftM ConstantIntRef.getTrue 
    let _ ← liftM (ConstantIntRef.ofNat 32 0)
    let _ ← liftM (ConstantIntRef.ofInt 64 (-1))
    let _ ← Builder.stringPtr "hello" AddressSpace.default true ""
    let x ← Builder.stringPtr "hello" 5 true ""
    let _ ← pure x 
    pure () : ModuleM PUnit
